From: Sankar P <psankar@novell.com>
Subject: [PATCH] fs: novfs: Fix exit handlers on local_unlink
References: bnc#649625
Patch-mainline: No

Unlock mutex and don't leak a dentry while exiting
the function local_unlink in case of invalid inode.

Reported-by: Jeremy Meldrum <jmeldrum@novell.com>
Signed-off-by: Tukaram Laxmeshwar <tlaxmeshwar@novell.com>
Signed-off-by: Michal Kubecek <mkubecek@suse.cz>
Signed-off-by: Sankar P <psankar@novell.com>
Acked-by: Jan Kara <jack@suse.cz>

diff --git a/fs/novfs/daemon.c b/fs/novfs/daemon.c
index 9ede03c..0fe24c3 100644
--- a/fs/novfs/daemon.c
+++ b/fs/novfs/daemon.c
@@ -1902,7 +1902,7 @@ static long local_unlink(const char *pathname)
 		if (!(dentry->d_inode) || !(dentry->d_inode->i_mode & S_IFLNK)) {
 			DbgPrint("%s not a link", name);
 			error = -ENOENT;
-			goto exit1;
+			goto exit2;
 		}
 		/* Why not before? Because we want correct error value */
 		if (nd.last.name[nd.last.len])
